sktime/transformations/base.py

```diff
@@ -48,93 +104,567 @@ class BaseTransformer(BaseEstimator):
+
+        # 3. internal only has Series but X is Panel: loop over instances
+        if (
+            X_input_scitype == "Panel"
+            and "Panel" not in X_inner_scitypes
+            and "Series" in X_inner_scitypes
+        ):
+            if y is not None:
+                ValueError(
+                    "no default behaviour if _fit does not support Panel, "
+                    " but X is Panel and y is not None"
+                )
+            X = convert_to(X, to_type="df-list", as_scitype="Panel")
+
+            if self.get_tag("fit-in-transform"):
+                Xt = [clone(self).transform(Xi) for Xi in X]
+            else:
+                transformers = self.transformers_
+                if len(transformers) != len(X):
+                    raise RuntimeError(
+                        "found different number of instances in transform than in fit"
+                    )
+                else:
+                    Xt = [transformers[i].transform(X[i]) for i in range(len(X))]
+            # now we have a list of transformed instances
+
+            # if the output is Series, Xt is a Panel and we convert back
+            if output_scitype == "Series":
+                Xt = convert_to(Xt, to_type=X_input_mtype, as_scitype="Panel")
+
+            # if the output is Primitives, we have a list of one-row dataframes
+            # we concatenate those and overwrite the index with that of X
+            elif output_scitype == "Primitives":
+                Xt = pd.concat(Xt)
+                Xt.index = X.index
+            return Xt
+
+        # convert X/y to supported inner type, if necessary
+        ###################################################
+
+        # variables for the scitype of the current X (possibly converted)
+        #     y wasn't converted so we can use y_input_scitype
+        X_mtype = mtype(X)
+        X_scitype = mtype_to_scitype(X_mtype)
+
+        # subset to the mtypes that are of the same scitype as X/y
+        X_inner_mtype = [
+            mt for mt in X_inner_mtype if mtype_to_scitype(mt) == X_scitype
+        ]
+
+        y_inner_mtype = [
+            mt for mt in y_inner_mtype if mtype_to_scitype(mt) == y_input_scitype
+        ]
+
+        # convert X and y to a supported internal type
+        #  if X/y type is already supported, no conversion takes place
+        X_inner = convert_to(
+            X,
+            to_type=X_inner_mtype,
+            as_scitype=X_scitype,
+        )
+        y_inner = convert_to(
+            y,
+            to_type=y_inner_mtype,
+            as_scitype=y_input_scitype,
+        )
+
+        # carry out the transformation
+        ###################################################
+
+        # todo: uncomment this once Z is completely gone
+        # Xt = self._transform(X=X_inner, y=y_inner)
+        # less robust workaround until then
+        Xt = self._transform(X_inner, y_inner)
+
+        # convert transformed X back to input mtype
```

Comment:
I don't agree with converting back to the input type. We should define a clear preferred machine type for a given scitype and use that convention consistently in sktime.

What happens in pipelines? Would we just convert to inner type, do transform and convert back to original type. Then do the same again for each step? That would lead to lots of unnecessary type conversions.

At a minimum I think we need to add a keyword argument that indicates whether the user wants to convert back to their input type (note that this would require us to have preferred machine types for given scitypes). It should default to not converting back to the input type (i.e. default to efficiency).

Reply:
Yes, I fully agree. That has been my main worry with this approach too. An additional downside is that it makes it hard to read and maintain different classes, because you would first have to understand what the machine type is. We need to limit the supported internal mtypes to a few key types like pandas Series/DataFrames and numpy arrays (see my comment above https://github.com/alan-turing-institute/sktime/pull/1365#discussion_r738185223)

Reply:
> I don't agree with converting back to the input type. We should define a clear preferred machine type for a given scitype and use that convention consistently in sktime.

Argh, this discussion truly is a constant annoyance... nothing personal, @RNKuhns or @mloning, but we need to make a decision.

In forecasting, we just reverted to "convert back to input type" from "convert to preferred type", didn't we, @aiwalter, @mloning? I'm really really happy with either as long as we consistently decide to do one.

>At a minimum I think we need to add a keyword argument that indicates whether the user wants to convert back to their input type (note that this would require us to have preferred machine types for given scitypes). It should default to not converting back to the input type (i.e. default to efficiency).

This would violate the unified interface principle, since transformers would no longer be interchangeable (they would randomly return outputs of different return mtypes). Disagreed, hence, since the unified interface principle is above minor efficiency considerations.

Reply:
I'm interpreting the "thumbs up" as agreement